Release/0.1.0

ninjabear opened the pull request.

alexanderdean commented:
Hey @ninjabear - this is coming along great! A few pieces of feedback, all about packaging really:

1. Instead of a Vagrantfile which just rust-ups, please use the existing Ansible rust playbook we have (if it needs updating, please go for it). So copy the Vagrantfile from https://github.com/snowplow/makefile-rs/blob/master/Vagrantfile and the vagrant subfolder here https://github.com/snowplow/makefile-rs/tree/master/vagrant
2. We want an easy way (pending CI/CD) of publishing new versions - so let's add a `vagrant push`, similar to the syntax in here: https://github.com/snowplow/sql-runner/blob/master/Vagrantfile#L20 and with a push.bash similar to: https://github.com/snowplow/sql-runner/blob/master/vagrant/push.bash
3. The README instructions - can we add a short Factfile example into the README, plus a link to a wiki page with various examples
4. The README has lots of instructions on how to build Factotum that are only really interesting to developers. Can we move those down into a Developer Quickstart section, and lead with a User Quickstart section, which has instructions on DLing from Bintray

Basically for 3 and 4 - it would be great for the README format to be very similar in flow to: https://github.com/snowplow/sql-runner/blob/master/README.md

alexanderdean commented:
Also @ninjabear - can you create a blog post PR please? Suggested contents:

1. Our use case for DAGs
2. Why create Factotum - why not Chronos et al, including Zen of Factotum
3. User quick start and example Factotum file - including importance of factfiles being JSON Schema based so easily validatable
4. Roadmap - what's coming next and what we want Factotum to evolve into

ninjabear commented:
Hey @alexanderdean, I think I've cycled through the same thought process as you (in fact my googling came up with [this](https://groups.google.com/forum/#!topic/vagrant-up/LgqE-JFAqZc))!

I agree with you that vagrant push will get us able to deploy factotum the fastest. I'll include that in this PR. However, thinking a bit ahead to CD the following things come up:

1. can travis run vagrant push (a VM in a VM?)
   * how would we get the keys in?
   * can't we run the provisioning on the travis box (if needed for the build)?
   * rust is statically linked and able to cross compile (the native isn't linked to any interpreter or lib 
version - the environment it's built/run on can be less strictly controlled)
        * having said that; I think the cross compilation actually involves a [fair bit of provisioning](https://github.com/japaric/rust-cross#requirements) (when we come to it)

2. having a script to push to bintray isn't quite the ["travis" way of doing it](https://docs.travis-ci.com/user/deployment/bintray)
    * this would mean putting a script in that writes the "descriptor" json out prior to deployment
    * this would result in two different scripts to deploy factotum (if we did it the travis way)

I'm searching around for a solid solution to this, to be honest I don't have a completely compelling answer yet - so apologies for the ramble but I thought I'd share my thoughts.

alexanderdean commented:
Hey @ninjabear, good stuff - let me know as soon as the push.bash is available as I will then push the rc1 and get this into test!

On your points:

> 1. can travis run vagrant push (a VM in a VM?)

It feels like we would jettison the Vagrant layer, and just provision into the Travis VM. We would need to come to a view on whether we keep the Ansible-driven provisioning (`oss-playbooks/rust.yml`) or not. The Ansible layer might step on existing software in the Travis VM, but if we get rid of it then our dev and build environments have diverged..

> 2. having a script to push to bintray isn't quite the "travis" way of doing it

The Travis documentation on how to push to Bintray is really interesting. Given that the descriptor.json contains version information in it, won't we need some kind of script to update that JSON? Or are we assuming it will be manually updated by the user before pushing to master? I think if we go the Travis way, it's fine to drop the current push.bash script as it stands today - i.e. we should only ever need one push strategy in a given repo...

ninjabear commented:
Hey @alexanderdean I've made the suggested readme alterations, added `vagrant push` and an initial changelog. 

I'll also start to get a blog post together! 

This is ready to try out - let me know if there's any problems with `vagrant push`. I tested it out as best I could without touching the snowplow bintray but some issues may have creeped in while doing the final edit.

alexanderdean commented:
Hey @ninjabear - that's awesome, I will push a version of this later today and give it a spin!

alexanderdean commented:
@ninjabear - can you rebase all of this down into one commit, "Initial commit"? We'll close the tickets in 0.1.0 by hand when the release comes out...

ninjabear commented:
Hey @alexanderdean - I've squashed this branch down into a single commit, but I think you might mean squash in any commits in master too - let me know if this is the case!

**I've also renamed 'steps' to tasks**

alexanderdean commented on README.md:
```diff
@@ -5,23 +10,100 @@
+* Compile and run a demo - `cargo run -- samples/echo.factotum` 
 
-In case of failure, we return a 1 error code and print out a summary of what succeeded and what failed to stderr.
+### Using stable rust without vagrant 
```
Rust not rust

alexanderdean commented on README.md:
```diff
@@ -5,23 +10,100 @@
+### Using stable rust without vagrant 
 
-In the case of a noop - needs discussion, #12.
+* **[Install rust](https://www.rust-lang.org/downloads.html)**
```
Rust not rust

alexanderdean commented on Vagrantfile:
```diff
@@ -0,0 +1,26 @@
+    vb.name = Dir.pwd().split("/")[-1] + "-" + Time.now.to_f.to_i.to_s
+    vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
+    vb.customize [ "guestproperty", "set", :id, "--timesync-threshold", 10000 ]
+    # Rust/Kotlin are not too memory-hungry
```
Remove Kotlin ref

alexanderdean commented on samples/echo.factotum:
```diff
@@ -0,0 +1,41 @@
+{
```
Can we change this directory from `samples/` to `examples/` please

ninjabear replied:
I had avoided `examples/` because these are normally reserved for Rust example code and can be executed with `cargo run --example xyz`

alexanderdean replied:
Ah! Okay cool, didn't know that

alexanderdean commented on samples/sleep.factotum:
```diff
@@ -0,0 +1,52 @@
+{
```
Should the file extension be `.factotum` or `.factfile`? Not sure

ninjabear replied:
This can change in a later release without too much fuss - I think we should create a ticket for it in a later milestone

alexanderdean replied:
Agree, sounds good

ninjabear replied:
raised https://github.com/snowplow/factotum/issues/34

alexanderdean commented on VERSION:
```diff
@@ -0,0 +1 @@
+0.1.0
```
Can we have this .gitignored and just automatically generated by parsing the Cargofile? This is what we do with all our Scala apps...

ninjabear replied:
Sure - I thought the file was there to ensure the tag tallies up with the version, I don't know if we need the file at all. If not I will .gitignore and generate it as you say

alexanderdean commented on CHANGELOG:
```diff
@@ -0,0 +1,3 @@
+Version 0.1.0 (2016-04-4)
```
Leave as `-xx)` until finalized

alexanderdean commented on .gitignore:
```diff
@@ -13,3 +13,6 @@
 # Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
 # More information here http://doc.crates.io/guide.html#cargotoml-vs-cargolock
 Cargo.lock
+.factotum
```
Update this if we update the fileext

alexanderdean commented:
Can we also rename `LICENSE` to `LICENSE-2.0.txt` for consistency with our other projects

alexanderdean commented on README.md:
````diff
@@ -5,23 +10,100 @@
-```bash
-$ factotum job run FILE --var env=prod --var email=bob@acme.com
+```{bash}
+wget http://dl.bintray.com/snowplow/snowplow-generic/factotum_0.1.0_linux.zip
````
The artifact that vagrant push is currently building is called `factotum_0.1.0-rc1_linux.zip`, note the hyphen not underscore; can we apply a `tr` to turn any -s in versions into _s for consistency (we do this in Schema Guru and other projects)...

alexanderdean replied:
In other apps we suffix with `_linux_amd64.zip` to indicate it's 64bit - can we do this here?

ninjabear replied:
RE: `s/-/_/g` that's cool will do that

RE: `_linux_amd64.zip`

Yes, we can do that - the reason I removed it is probably just me being pedantic. Other than very likely being 64 bit, I couldn't tell you if it's amd64/intel64*. It's probably some code that falls in both groups which tends to be called x86_64 (amd64 and intel64 are nearly identical, to the point amd64 sometimes refers to both). Since the default to me is x86_64 (i686/32bit would need specifying) I removed it altogether.

.* intel64 isn't its actual name

alexanderdean replied:
I hear you - I copied `_linux_amd64` from what HashiCorp do... I am happy with `_linux_amd64` or ``_linux_x86_64` as you prefer, but would rather over-specify than under-specify...

ninjabear replied:
Sure will add it in!

alexanderdean commented:
CI/CD bot says:

> Artifact published to Bintray! https://bintray.com/artifact/download/snowplow/snowplow-generic/factotum_0.1.0-rc1_linux.zip

alexanderdean commented:
@ninjabear I am getting an error with the rc1 I've built and deployed for testing:

```
$ wget https://raw.githubusercontent.com/snowplow/factotum/release/0.1.0/samples/noop.factotum
$ /opt/factotum-0.1.0-rc1/factotum noop.factotum
No such file or directory (os error 2)
Executing task 'echo alpha'!
Task 'echo alpha' wrote the following to STDOUT:
alpha
Task 'echo alpha' after PT0.002303591S returned 0
Executing task 'exit'!
Job will now stop as task 'exit' ended with 99
```

Happens with any factfile... My gut says that Factotum is trying to load some internal resource which isn't being found...

ninjabear commented:
Hey @alexanderdean, you're quite right - it's the log file config! It's looking for `./config/log.toml`

ninjabear commented:
Hey @alexanderdean - can you try an `rc2` release please!

# What's changed

* fixed references to "rust" (no uppercase) in readme
* removed Kotlin reference in Vagrantfile
* renamed LICENSE to LICENSE-2.0 (although the commit message for https://github.com/snowplow/factotum/pull/32/commits/43549031c2701143b4b34b259e4e28946ae28817 is the UK spelling, whoops)
* replaced the release date with `xx` in the changelog
* replaced - in 0.1.0-rc1 with an underscore when naming artifacts
* added in the arch as x86_64 for the artifact
* fixed that error on loads without a `config/log.toml`
    * I fixed this by removing the external configuration (no `config/log.toml` anymore) and using an inbuilt config builder 
* When entering the API key in push.bash, the key is no longer echo'd to stdout (how sudo works)

I will squash these commits once 0.1.0-rc2 has built successfully!

alexanderdean commented:
@ninjabear I'm now getting:

```
UPLOADING ARTIFACT TO BINTRAY*
* ~2 minutes
-------------------------------
curl: Can't open './target/release/factotum_0.1.0_rc2_linux_x86_64.zip'!
curl: try 'curl --help' or 'curl --manual' for more information
...
file ./target/release/factotum_0.1.0_rc2_linux_x86_64.zip
./target/release/factotum_0.1.0_rc2_linux_x86_64.zip: Zip archive data, at least v2.0 to extract
```

Any ideas?

ninjabear commented:
The working directory wasn't properly reset before the curl!

alexanderdean commented:
```perl
perl -ne 'print $1 if m/version\s*=\s*\"(\d+\.\d+\.\d+)\"/' Cargo.toml
```

doesn't work with `0.1.0-rc2`

ninjabear commented:
Oh! I was expecting you to set the version in `Cargo.toml` to the "real" version, and ensure the tag is `0.1.0-rc1`!

```{bash}
 # take the latest tag, stripping anything not looking like a version (e.g. 0.1.0-rc99 => 0.1.0)
 tag_version=`git describe --abbrev=0 --tags | perl -ne 'print $1 if /(\d+\.\d+\.\d+)/'` 
```

alexanderdean commented:
Sorry, the version is still not quite right in Bintray:

https://bintray.com/snowplow/snowplow-generic/factotum/view

The version should be 0.1.0-rc2, not underscore. (Underscores only for artifact filenames)

ninjabear commented:
I see what you mean! The artifact name now has underscores, but the artifact version on bintray does not, here's my test version: 

https://bintray.com/ed/generic-test/factotum/view

alexanderdean commented:
Confirmed! Thanks so much
